### DISCUSSION TITLE
Selector for useState

### DISCUSSION THREAD

**Discussion by doliG** [pmndrs/zustand#1484] [General]
Hello,

Is it considered as good practice to create custom selector function for the store. I.E:

```js
const store = create((set) => ({
  mode: "add", // can be "add" | "remove" | "create"
  input: 10,
  balance: 5,
  
  setInput: (input) => set({ input }),
  fetchBalance: () => /* ... */
}))

const selectResult = (state) => {
  const { input, balance } = state
  switch (s.mode) {
    case "add": return balance + input
    case "remove": return balance - input
    case "create": return input
  }
}

function Result() {
  const res = useStore(selectResult)

  return <p>{res}</p>
}
```

The problem I'm trying to solve here is to avoid having tons of ternary in the template: with the selector I can grab directly the value I need to display.

I searched the docs and found only a few lines where this pattern is discussed: 
> If a selector doesn't depend on scope, you can define it outside the render function to obtain a fixed reference without useCallback.
> 
> const selector = state => state.berries
> 
> function Component() {
>   const berries = useStore(selector)
> 
> https://github.com/pmndrs/zustand/blob/main/docs/recipes/recipes.mdx#memoizing-selectors

If you have other ideas feels free to tell me. Thanks :)

**Comment by dai-shi** [comment]
> Is it considered as good practice to create custom selector function for the store. 

Yes, it's good practice.

**Reply by doliG** [reply]
Hey, thanks for your blazing fast answer (as usual haha). Also is it fine to make some light computation in the selectors, i.e computing some values ? 🤔

**Reply by dai-shi** [reply]
sure. if you have a heavy computation, you should consider memoizing selectors.

**Reply by doliG** [reply]
well I don't think that's possible because the deps are all in the store, or I need to do some hacky like

```js
const compute = (a, b) => a + b

const selector = (state) => {
  const { a, b } = state;
  return useMemo(compute(a, b), [a, b])
}
```

Which breaks the composability, because I won't be able to use them in other store cause useMemo is only allowed into components 🤔

**Reply by dai-shi** [reply]
You might find [proxy-memoize](https://github.com/dai-shi/proxy-memoize) interesting.
